cpp/src/cuda_event.cpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <rapidsmpf/cuda_event.hpp>
+
+namespace rapidsmpf {
+
+CudaEvent::CudaEvent(unsigned flags) {
+    RAPIDSMPF_CUDA_TRY(cudaEventCreateWithFlags(&event_, flags));
+}
+
+std::shared_ptr<CudaEvent> CudaEvent::make_shared_record(
+    rmm::cuda_stream_view stream, unsigned flags
+) {
+    auto ret = std::make_shared<CudaEvent>(flags);
+    ret->record(stream);
+    return ret;
+}
+
+CudaEvent::~CudaEvent() noexcept {
+    if (event_ != nullptr) {
+        cudaEventDestroy(event_);
```

Comment:
Just to note in passing, we're not allowed to make `static`-lifetime `CudaEvent`s